websockets/README.md

```diff
@@ -50,9 +50,27 @@ WebSockets have no built in authentication mechanism. Server-side processes list
 
 ## Encryption
 
-At the time of writing, the negotiated authentication mechanism should also be used to encrypt all traffic sent over the WebSocket even if TLS certificates are also used at the transport layer.
+Server-side processes listening on WebSocket addresses should use TLS certificates to secure transmitted data at the transport level.
 
-A mechanism to avoid this but also maintain backwards compatibility with existing server-side processes will be specified in a future revision to this spec.
+This does not provide any assurance that the remote peer possesses the private key that corresponds to their public key, so an additional handshake is necessary.
+
+During connection establishment over WebSockets, before the connection is made available to the rest of the application, if all of the following criteria are met:
+
+1. `noise` is negotiated as the connection encryption protocol
+2. An initial handshake is performed with the `handshake_only` boolean extension set to true
+3. The transport layer is secured by TLS
+
+Then all subsequent data is sent without encrypting it at the libp2p level, instead relying on TLS encryption at the transport layer.
+
+If any of the above is not true, all data is encrypted with the negotiated connection encryption method before sending.
+
+This prevents double-encryption but only when both ends opt-in to ensure backwards compatibility with existing deployments.
+
+### MITM mitigation
```

Comment:
This doesn't work unless the server somehow includes their domain name in their noise handshake. Otherwise, an attacker can simply announce that some peer id `QmVictim` is available at `/dns/attacker.com/...`.

Furthermore, by including the domain name in the noise handshake, the server is _opting-in_ to trusting the CA system.

Reply:
Note that this also doesn't work unless the server trusts all owners of the given domain name. If the server operator is the sole owner of the domain name, this is fine. But if another party controls the domain name as well, but not the private key to the peer ID, they would be able to MITM/intercept this connection.

Reply:
Good points both.

> This doesn't work unless the server somehow includes their domain name in their noise handshake

I've added a further `tls_common_name` extension to the noise handshake to prevent an attacker publishing an address with a bogus domain and then simply relaying the handshake.

> by including the domain name in the noise handshake, the server is opting-in to trusting the CA system

I've added a note to this effect.

> Note that this also doesn't work unless the server trusts all owners of the given domain name

I've copied the `Security Considerations` section from #564 as it seems like they have the same problems with third party root certificates installed on the machine.

Do you think this covers it?